src/centralized_telescope.rs

```diff
@@ -0,0 +1,419 @@
+    /// Oracle producing a uniformly random value in [1, n_p] used for round candidates
+    /// We also return hash(data) to follow the optimization presented in Section 3.3
+    fn h1(input: &[Vec<u8>], n_p: u64) -> (Hash, u64) {
+        let mut data = vec!["Telescope-H1".as_bytes().to_vec()];
+        for i in input {
+            data.push(i.to_vec());
+        }
+        let digest = utils::combine_hashes::<DIGEST_SIZE>(&data);
+        (digest, utils::sample_uniform(&digest, n_p))
+    }
+
+    /// Output a round from a proof counter and n_p
+    /// Initilialises the hash with H1(t) and random value as oracle(H1(t), n_p)
+    pub fn new(v: u64, t: u64, n_p: u64) -> Round {
+        let mut data = vec![v.to_ne_bytes().to_vec()];
+        data.push(t.to_ne_bytes().to_vec());
+        let (h, h_u64) = Round::h1(&data, n_p);
+        Round {
+            v,
+            t,
+            s_list: vec![],
+            h,
+            h_u64,
+            n_p,
+        }
+    }
+
+    /// Updates a round with an element of S_p
+    /// Replaces the hash $h$ with $h' = H1(h, s)$ and the random value as oracle(h', n_p)
+    pub fn update(r: &Round, s: Element) -> Round {
+        let mut s_list = r.s_list.clone();
+        s_list.push(s);
+        let mut data = vec![r.h.clone().to_vec()];
+        data.push(s.to_vec());
+        let (h, h_u64) = Round::h1(&data, r.n_p);
+        Round {
+            v: r.v,
+            t: r.t,
+            s_list,
+            h,
+            h_u64,
+            n_p: r.n_p,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+/// Alba proof
+pub struct Proof {
+    /// Proof counter
+    v: u64,
+    /// Proof 2nd counter
+    t: u64,
+    /// Proof tuple
+    items: Vec<Element>,
+}
+
+impl Proof {
+    /// Oracle producing a uniformly random value in [1, n_p] used for prehashing S_p
+    fn h0(setup: &Setup, v: u64, s: Element) -> u64 {
+        let mut data = vec!["Telescope-H0".as_bytes().to_vec()];
+        data.push(v.to_ne_bytes().to_vec());
+        data.push(s.to_vec());
+        let digest = utils::combine_hashes::<DIGEST_SIZE>(&data);
+        utils::sample_uniform(&digest, setup.n_p)
+    }
+
+    /// Oracle defined as Bernoulli(q) returning 1 with probability q and 0 otherwise
+    fn h2(setup: &Setup, r: &Round) -> bool {
+        let mut data = vec!["Telescope-H2".as_bytes().to_vec()];
+        data.push(r.h.to_vec());
+        let digest = utils::combine_hashes::<DIGEST_SIZE>(&data);
+        utils::sample_bernouilli(&digest, setup.q)
+    }
+
+    /// Depth-first search which goes through all potential round candidates
+    /// and returns first round candidate Round{t, x_1, ..., x_u)} such that:
+    /// - for all i ∈ [0, u-1], H0(x_i+1) ∈ bins[H1(t, x_1, ..., x_i)]
+    /// - H2(t, x_0, ..., x_u) = true
+    fn dfs(
+        setup: &Setup,
+        bins: &[Vec<Element>],
+        round: &Round,
+        limit: u64,
+    ) -> (u64, Option<Proof>) {
+        if round.s_list.len() as u64 == setup.u {
```

Comment:
lines 279--288: apparently Rust allows you to use a return statement with ternary operator
```
return if Proof::h2(setup, round) {
    (limit, Some(Proof {
         v: round.v,
          t: round.t,
          items: round.s_list.clone(),
     }))
} else {
    (limit, None)
};
```

Reply:
I saw. I first wrote somethine like the following but it wasn't wasy to read 
```return (limit, if Proof::h2(setup, round) {
    (limit, Some(Proof {
         v: round.v,
          t: round.t,
          items: round.s_list.clone(),
     })) else { None })
};```
I'll update with yours!

Reply:
Another option:
```
let proof = if Proof::h2(setup, round) {
    Some(Proof {
         v: round.v,
          t: round.t,
          items: round.s_list.clone(),
     })
} else {
    None
};
return (limit, proof)
```